portal/server/utils/csv-utils.js

```diff
@@ -0,0 +1,138 @@
+const ExcelJS = require('exceljs');
+const csv = require('csv-parser');
+const { Readable } = require('stream');
+const { CELL_ADDRESS_REGEX } = require('../constants/regex');
+
+const columnIndexToExcelColumn = (index) => {
+  let result = '';
+  let indexTracker = index;
+  while (indexTracker >= 0) {
+    result = String.fromCharCode((indexTracker % 26) + 65) + result;
+    indexTracker = Math.floor(indexTracker / 26) - 1;
+  }
+  return result;
+};
+
+const parseXlsxToCsvArrays = (worksheet) => {
+  const csvData = [];
+  const csvDataWithCellAddresses = [];
+
+  let firstRow = true;
+  worksheet.eachRow((row) => {
+    const rowData = [];
+    const rowDataWithCellAddresses = [];
+    row.eachCell({ includeEmpty: true }, (cell) => {
+      rowData.push(cell.value);
+      if (firstRow) {
+        rowDataWithCellAddresses.push(`${cell.value}`);
+      } else {
+        rowDataWithCellAddresses.push(`${cell.value}-${cell.address}`);
+      }
+    });
+    firstRow = false;
+    csvData.push(rowData.join(','));
+    csvDataWithCellAddresses.push(rowDataWithCellAddresses.join(','));
+  });
+  return { csvData, csvDataWithCellAddresses };
+};
+
+const xlsxBasedCsvToJsonPromise = async (csvDataWithCellAddresses) => {
+  const csvStream = new Readable({
+    read() {
+      for (const line of csvDataWithCellAddresses) {
+        this.push(`${line}\n`); // Add newline to simulate line-by-line reading
+      }
+      this.push(null); // End the stream
+    },
+  });
+  return new Promise((resolve, reject) => {
+    try {
+      const jsonParsedData = [];
+      csvStream
+        .pipe(
+          csv({
+            mapHeaders: ({ header }) => header.toLowerCase().replace(/\s/g, ' ').trim(),
+            mapValues: ({ value }) => {
+              const lastDashIndex = value.lastIndexOf('-');
+              const cellValue = value.substring(0, lastDashIndex);
+              const cellAddress = value.substring(lastDashIndex + 1);
+
+              // Split the cell address into column and row
+              const [column, row] = cellAddress.match(CELL_ADDRESS_REGEX).slice(1);
+              return { value: cellValue !== 'null' ? cellValue : null, column, row };
+            },
+          }),
+        )
+        .on('data', (row) => {
+          jsonParsedData.push(row);
+        })
+        .on('end', () => {
+          resolve(jsonParsedData);
+        });
+    } catch (error) {
+      console.error(error);
+      reject(new Error('Error parsing CSV data'));
+    }
+  });
+};
+
+const csvBasedCsvToJsonPromise = async (csvBuffer) => {
+  const csvStream = Readable.from(csvBuffer);
+  return new Promise((resolve, reject) => {
+    try {
+      const csvData = [];
+      csvStream
+        .pipe(
+          csv({
+            mapHeaders: ({ header }) => header.toLowerCase().replace(/\s/g, ' ').trim(),
+            mapValues: ({ index, value }) => ({ value, column: columnIndexToExcelColumn(index), row: null }),
```

Comment:
I'm curious why the row values are `null` here, but it might make sense when I read the rest of the code

Reply:
Okay my understanding from looking at the code is:
- the validators want to know which row caused the validation to fail, so we can tell the user where the error is
- `csv-parser` doesn't give us the row number in mapValues, so we can't work out the row number here and set it to `null` instead
- when the validators validate the cell, we pass in the `rowNumber` which we've worked out from the JSON array index instead, which lines up with the actual CSV row number
- for XLSX based CSVs, the JSON array index does **not** match the actual row number because XSLXs can have empty rows, so in that case we have to use `row` instead

Is that right? It's a bit tricky to work out from the code if so, is there any way we can simplify it, e.g. can weset the row number of the csv based JSONs in `csvBasedCsvToJsonPromise` somehow instead of leaving them `null` and having to work out the row number later?

Reply:
You're pretty much correct in everything you've said. The only way I can think of getting row number in the csvBasedCsvToJsonPromise function is to iterate through the array and add it on in a separate loop (which is what I do later on to get the row number anyway). This would possibly be a bit neater/easier to read at the cost of a bit of performance (that should be completely negligible in this case). I can give that a go and see if it works out nicely!